File: 79.word-search.py
#
# @lc app=leetcode id=79 lang=python3
#
# [79] Word Search
#

# @lc code=start
from collections import Counter
from itertools import chain, product, starmap


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n, m, wrdlen = len(board), len(board[0]), len(word)
        if wrdlen > m * n:
            return False

        def dfs(r, c, i=0) -> bool:
            if board[r][c] != word[i]:
                return False
            if i == wrdlen - 1:
                return True
            
            board[r][c] = "#"
            i += 1
            valid = (
                (r > 0 and dfs(r - 1, c, i))
                or (r < n - 1 and dfs(r + 1, c, i))
                or (c > 0 and dfs(r, c - 1, i))
                or (c < m - 1 and dfs(r, c + 1, i))
            )
            board[r][c] = word[i - 1]
            return valid

        wordcnt, boardcnt = Counter(word), Counter(chain(*board))
        if any(wordcnt[c] > boardcnt[c] for c in wordcnt):
            return False
        if wordcnt[word[0]] > wordcnt[word[-1]]:
            word = word[::-1]
        return any(starmap(dfs, product(range(n), range(m))))

# @lc code=end

File: RustSolutions/src/79.word-search.rs
/*
 * @lc app=leetcode id=79 lang=rust
 *
 * [79] Word Search
 */

// @lc code=start
impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let mut board_cpy = board.clone();
        for i in 0..board.len() {
            for j in 0..board[0].len() {
                if Self::dfs(&mut board_cpy, &word, i, j, 0) {
                    return true;
                }
            }
        }
        return false;
    }

    pub fn dfs(board: &mut Vec<Vec<char>>, word: &String, i: usize, j: usize, k: usize) -> bool {
        if
        i<0 || j<0 ||
        i >= board.len() || j >= board[0].len() ||
        board[i][j] != word.chars().nth(k).unwrap()
        {
            return false;
        }
        if k == word.len() - 1 {
            return true;
        }
        board[i][j] = '#';
        let res =
            Self::dfs(board, word, i + 1, j, k + 1)
            || Self::dfs(board, word, i - 1, j, k + 1)
            || Self::dfs(board, word, i, j + 1, k + 1)
            || Self::dfs(board, word, i, j - 1, k + 1);
        board[i][j] = word.chars().nth(k).unwrap();
        res
    }
}
// @lc code=end

